File: ios/VidyoConnector/VidyoConnectorViewManager.m
//
//  VidyoConnectorViewManager.m
//  VidyoReactNative
//
//  Created by serhii benedyshyn on 4/21/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "VidyoConnectorView.h"
#import "VidyoConnectorViewManager.h"
#import <UIKit/UIKit.h>
#import <MapKit/MapKit.h>
#import <React/RCTBridge.h>
#import <React/RCTViewManager.h>
#import <React/RCTUIManager.h>

@implementation VidyoConnectorViewManager : RCTViewManager

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE()

RCT_CUSTOM_VIEW_PROPERTY(viewStyle, NSString*, VidyoConnectorView) {
  [view setViewStyle:[[RCTConvert NSString:json] isEqual: @"ViewStyleTiles"] ? VCConnectorViewStyleTiles : VCConnectorViewStyleDefault];
}
RCT_CUSTOM_VIEW_PROPERTY(remoteParticipants, NSString*, VidyoConnectorView) {
  [view setRemoteParticipants:[RCTConvert int:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(logFileFilter, NSString*, VidyoConnectorView) {
  [view setLogFileFilter:[RCTConvert NSString:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(logFileName, NSString*, VidyoConnectorView) {
  [view setLogFileName:[RCTConvert NSString:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(userData, NSString*, VidyoConnectorView) {
  [view setUserData:[RCTConvert int:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(cameraPrivacy, NSString*, VidyoConnectorView) {
  [view setCameraPrivacy:[RCTConvert BOOL:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(microphonePrivacy, NSString*, VidyoConnectorView) {
  [view setMicrophonePrivacy:[RCTConvert BOOL:json]];
}
RCT_CUSTOM_VIEW_PROPERTY(mode, NSString*, VidyoConnectorView) {
  [view setMode:[[RCTConvert NSString:json] isEqual: @"VIDYO_CONNECTORMODE_Background"] ? VCConnectorModeBackground : VCConnectorModeForeground];
}

RCT_EXPORT_VIEW_PROPERTY(onConnect,                   RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDisconnect,                RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onParticipantJoined,         RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onParticipantLeft,           RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onDynamicParticipantChanged, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onLoudestParticipantChanged, RCTBubblingEventBlock)

RCT_EXPORT_METHOD(connect:(nonnull NSNumber*)reactTag
                  Portal:(NSString*)portal
                  RoomKey:(NSString*)roomKey
                  RoomPin:(NSString*)roomPin
                  DisplayName:(NSString*)displayName)
{
  [self.bridge.uiManager addUIBlock:
   ^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry){
    VidyoConnectorView *view = viewRegistry[reactTag];
    if (![view isKindOfClass:[VidyoConnectorView class]]) {
      RCTLogError(@"Invalid view returned from registry, expecting VidyoConnectorView, got: %@", view);
    }
    
    [view connectToRoomAsGuest:portal RoomKey:roomKey RoomPin:roomPin DisplayName:displayName];
  }];
}

RCT_EXPORT_METHOD(disconnect:(nonnull NSNumber*)reactTag)
{
  [self.bridge.uiManager addUIBlock:
   ^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry){
    VidyoConnectorView *view = viewRegistry[reactTag];
    if (![view isKindOfClass:[VidyoConnectorView class]]) {
      RCTLogError(@"Invalid view returned from registry, expecting VidyoConnectorView, got: %@", view);
    }
    [view disconnect];
  }];
}

RCT_EXPORT_METHOD(cycleCamera:(nonnull NSNumber*)reactTag)
{
  [self.bridge.uiManager addUIBlock:
   ^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, UIView *> *viewRegistry){
    VidyoConnectorView *view = viewRegistry[reactTag];
    if (![view isKindOfClass:[VidyoConnectorView class]]) {
      RCTLogError(@"Invalid view returned from registry, expecting VidyoConnectorView, got: %@", view);
    }
    [view cycleCamera];
  }];
}

- (UIView *)view
{
  if (vidyoView == nil) {
    vidyoView = [[VidyoConnectorView alloc] initWithBridge:self.bridge];
  }
  
  return vidyoView;
}

@end

File: android/app/src/main/java/com/vidyo/connector/view/VidyoConnectorViewManager.java
package com.vidyo.connector.view;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

import javax.annotation.Nullable;

public class VidyoConnectorViewManager extends SimpleViewManager<VidyoConnectorView> {

    public static final String REACT_CLASS = "VidyoConnectorView";

    private VidyoConnectorView vidyoConnectorView;
    private ThemedReactContext context;

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    public VidyoConnectorView createViewInstance(@NonNull ThemedReactContext context) {
        /* Refresh */
        if (this.context != context) {
            this.context = context;

            if (vidyoConnectorView != null) vidyoConnectorView.dispose();
            vidyoConnectorView = new VidyoConnectorView(context);
        }

        return vidyoConnectorView;
    }

    @ReactProp(name = "viewStyle")
    public void setViewStyle(VidyoConnectorView vc, String viewStyle) {
        vc.setViewStyle(viewStyle);
    }

    @ReactProp(name = "remoteParticipants")
    public void setRemoteParticipants(VidyoConnectorView vc, int remoteParticipants) {
        vc.setRemoteParticipants(remoteParticipants);
    }

    @ReactProp(name = "logFileFilter")
    public void setLogFileFilter(VidyoConnectorView vc, String logFileFilter) {
        vc.setLogFileFilter(logFileFilter);
    }

    @ReactProp(name = "logFileName")
    public void setLogFileName(VidyoConnectorView vc, String logFileName) {
        vc.setLogFileName(logFileName);
    }

    @ReactProp(name = "userData")
    public void setUserData(VidyoConnectorView vc, double userData) {
        vc.setUserData((long) userData);
    }

    @ReactProp(name = "cameraPrivacy")
    public void setCameraPrivacy(VidyoConnectorView vc, boolean privacy) {
        vc.setCameraPrivacy(privacy);
    }

    @ReactProp(name = "microphonePrivacy")
    public void setMicrophonePrivacy(VidyoConnectorView vc, boolean privacy) {
        vc.setMicrophonePrivacy(privacy);
    }

    @ReactProp(name = "mode")
    public void setMode(VidyoConnectorView vc, String mode) {
        vc.setMode(mode);
    }

    @Nullable
    @Override
    public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
        MapBuilder.Builder<String, Object> builder = MapBuilder.builder();

        return builder
                .put("onConnect", MapBuilder.of("registrationName", "onConnect"))
                .put("onFailure", MapBuilder.of("registrationName", "onFailure"))
                .put("onDisconnect", MapBuilder.of("registrationName", "onDisconnect"))

                .put("onParticipantJoined", MapBuilder.of("registrationName", "onParticipantJoined"))
                .put("onParticipantLeft", MapBuilder.of("registrationName", "onParticipantLeft"))
                .put("onDynamicParticipantChanged", MapBuilder.of("registrationName", "onDynamicParticipantChanged"))
                .put("onLoudestParticipantChanged", MapBuilder.of("registrationName", "onLoudestParticipantChanged"))

                .build();
    }

    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        MapBuilder.Builder<String, Integer> builder = MapBuilder.builder();

        return builder
                .put("connect", VidyoConnectorCommands.CONNECT.getCode())
                .put("disconnect", VidyoConnectorCommands.DISCONNECT.getCode())
                .put("cycleCamera", VidyoConnectorCommands.CYCLE_CAMERA.getCode())
                .build();
    }

    @Override
    public void receiveCommand(@NonNull VidyoConnectorView vc, int commandId, @Nullable ReadableArray params) {
        super.receiveCommand(vc, commandId, params);

        VidyoConnectorCommands command = VidyoConnectorCommands.match(commandId);

        switch (command) {
            case CONNECT:
                vc.connect(params);
                break;
            case DISCONNECT:
                vc.disconnect();
                break;
            case CYCLE_CAMERA:
                vc.cycleCamera();
                break;
        }
    }
}